File: src/Java/algorithms/mathandprob/StringToInteger.java
package algorithms.mathandprob; 

import java.util.*; 

public class StringToInteger {
    /* Function: Given a string, convert it into integer. The previous space should be omit. Stop 
     *     if encounter a none number character. If overflow 32 bit, return Integer.MAX/MIN_VALUE; 
     * Solution: When previous value is bigger than Integer.MAX_VALUE / 10 or equals it and current
     *     char is larger than Integer.MAX_VALUE % 10, can have 3 conditions: 1. just equals Integer.MAX; 
     *     2. larger than Integer.MAX; 3. equals to Integer.MIN_VALUE; 
     */ 
    
    public int atoi(String str) {
        if (str == null || str.length() == 0) {
            return 0; 
        }
        
        str = str.trim(); 
        int sign = str.charAt(0) == '-' ? -1 : 1; 
        if (str.charAt(0) == '+' || str.charAt(0) == '-') {
            str = str.substring(1); 
        }
        
        int sum = 0, pt = 0; 
        while (pt < str.length() && str.charAt(pt) >= '0' && str.charAt(pt) <= '9') {
            if (sum > Integer.MAX_VALUE / 10 || sum >= Integer.MAX_VALUE / 10 && str.charAt(pt) > Integer.MAX_VALUE % 10 + '0') {
                return sign > 0 ? Integer.MAX_VALUE : Integer.MIN_VALUE; 
            }
            sum = sum * 10 + str.charAt(pt++) - '0'; 
        }
        
        return sign * sum; 
    }
}

File: src/CPP/src/leetcode/StringToInteger.cpp
/*
 * [Source] https://leetcode.com/problems/string-to-integer-atoi/
 * [Difficulty]: Easy
 * [Tag]: Math
 * [Tag]: String
 */

#include <iostream>

using namespace std;

// [Solution]: Work backwards. So that if the number is overflowed, we can know it in advance.
// [Corner Case]: sign, overflow, white spaces at the begining, special characters.
class Solution {
};

// [Solution]: Work frontwards. If the previous value is larger than max_int / 10, it means overflow. If it equals max_int / 10, need consider the last digit.
/* Java solution
https://github.com/ZhenyingZhu/ClassicAlgorithms/blob/master/src/algorithms/mathandprob/StringToInteger.java
 */

/* Java solution
public class Solution {
    public int atoi(String str) {
        if (str == null || str.length() == 0) {
            return 0; 
        }
        int sign = 1, num = 0; 
        int i = 0; 
        while (str.charAt(i) == ' ' && i < str.length()) {
            ++i; 
        }
        if (str.charAt(i) == '+') {
            ++i; 
        } else if (str.charAt(i) == '-') {
            sign = -1; 
            ++i; 
        }
        for (; i < str.length(); ++i) {
            if (!Character.isDigit(str.charAt(i))) {
                break; 
            }
            if (num > Integer.MAX_VALUE / 10 || (num == Integer.MAX_VALUE / 10 && str.charAt(i) - '0' > Integer.MAX_VALUE % 10)) { // Overflow
                return sign == -1 ? Integer.MIN_VALUE : Integer.MAX_VALUE; 
            }
            num = num * 10 + str.charAt(i) - '0'; 
        }
        return sign * num; 
    }
}
 */

int main() {
    Solution sol;

    return 0;
}
